sdk/src/main/scala/com/horizen/account/api/rpc/service/EthService.scala

```diff
@@ -1,112 +1,178 @@
 
+  private def getBlockHeight = nodeView.history.getCurrentHeight
+
+  private def getStateDbRootFromBlockById(blockId: String) : AccountStateView = {
+    val block = nodeView.history.getBlockById(blockId)
+    if (block.isEmpty) return null
+    nodeView.state.getStateDbViewFromRoot(block.get().header.stateRoot)
+  }
+
   @RpcMethod("eth_chainId") def chainId = new Quantity(Numeric.toHexStringWithPrefix(BigInteger.valueOf(networkParams.chainId)))
 
   @RpcMethod("eth_getBalance") def GetBalance(address: String, blockNumberOrTag: Quantity) = new Quantity(getBalance(address, blockNumberOrTag))
 
+  private def getBlockIdByTag(tag: String): String = {
+    var blockId: java.util.Optional[String] = null
+    val history = nodeView.history
+    tag match {
+      case "earliest" => blockId = nodeView.history.getBlockIdByHeight(1)
+      //case "finalized" => //-39001: Unkown block
+      //case "safe" => //-39001: Unkown block
+      case "latest" => blockId = history.getBlockIdByHeight(nodeView.history.getCurrentHeight)
+      case "pending" => blockId = history.getBlockIdByHeight(nodeView.history.getCurrentHeight)
+      case _ => blockId = history.getBlockIdByHeight(Integer.parseInt(Numeric.cleanHexPrefix(tag), 16))
+    }
+    if (blockId.isEmpty) return null
+    blockId.get()
+  }
+
+  private def getBalance(address: String, tag: Quantity) : String = {
+    val stateDbRoot = getStateDbRootFromBlockById(getBlockIdByTag(tag.getValue))
+    if (stateDbRoot == null) return "null"
+    val balance = stateDbRoot.getBalance(Numeric.hexStringToByteArray(address))
+    Numeric.toHexStringWithPrefix(balance)
+  }
+
   @RpcMethod("eth_getTransactionCount") def getTransactionCount(address: String, tag: Quantity): Quantity = {
-    val nonce = nodeView.state.getNonce(Numeric.hexStringToByteArray(address))
+    val stateDbRoot = getStateDbRootFromBlockById(getBlockIdByTag(tag.getValue))
+    if (stateDbRoot == null) return new Quantity("null")
+    val nonce = stateDbRoot.getNonce(Numeric.hexStringToByteArray(address))
     new Quantity(Numeric.toHexStringWithPrefix(nonce))
   }
 
-  @RpcMethod("net_version") def version: String = Numeric.toHexStringWithPrefix(BigInteger.TEN) // TODO: sidechainSettings.genesisData.mcNetwork or just a number????
+  @RpcMethod("net_version") def version: String = String.valueOf(networkParams.chainId)
 
-  @RpcMethod("eth_estimateGas") def estimateGas(transaction: util.LinkedHashMap[String, String]/*, tag: Quantity*/) = { // TODO: We need an estimateGas function to execute EVM and get the gasUsed
-    if(transaction.containsKey("data") && transaction.get("data") != null && transaction.get("data").length > 1)
+  @RpcMethod("eth_estimateGas") def estimateGas(transaction: util.LinkedHashMap[String, String] /*, tag: Quantity*/) = { // TODO: We need an estimateGas function to execute EVM and get the gasUsed
```

Comment:
I think this method is important to have now, should call `eth_call` inside and get total gas used from it.